Add isclose() method to Time and TimeDelta classes

taldcroft opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

This pull request is to add an `isclose()` method to the `Time` and `TimeDelta` classes to make it easier to compare times to within a specified tolerance. The default tolerance is basically 1-bit in the mantissa of a float64 (`np.finfo(float).eps` = 2.2e-16 days or about 20 picosec), but the `atol` argument can be provided to be less strict.

This partially addresses #8742.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

pep8speaks commented:
Hello @taldcroft :wave:! Thanks for opening this pull request, we are very grateful for your contribution! I'm a friendly :robot: that checks for style issues in this pull request, since this project follows the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guidelines. I've listed some small issues I found below, but please don't hesitate to ask if any of them are unclear!




* In the file [`astropy/time/tests/test_comparisons.py`](https://github.com/astropy/astropy/blob/dc272471739177877efab98e7f0864d58049ac7e/astropy/time/tests/test_comparisons.py):

> [Line 124:1](https://github.com/astropy/astropy/blob/dc272471739177877efab98e7f0864d58049ac7e/astropy/time/tests/test_comparisons.py#L124): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

pllim commented:
Hmm, I guess `rtol` doesn't make sense in this context. Does this needs a "what's new" entry?

mhvk commented on astropy/time/core.py:
```diff
@@ -1180,6 +1180,34 @@ def __setitem__(self, item, value):
+            raise TypeError("'atol' argument must be a Quantity or TimeDelta instance, got "
+                            f'{atol.__class__.__name__} instead')
+
+        # Note: np.abs(self - other) results in an object array of TimeDelta
```
There should be no need to do that - `TimeDelta` should be smart enough! (And I think it is - if not, that's a bug...)

taldcroft replied:
See #10653

mhvk replied:
As noted there, plain `abs(...)` should work.

mhvk commented on astropy/time/core.py:
```diff
@@ -1180,6 +1180,34 @@ def __setitem__(self, item, value):
+            Default is one bit in the 128-bit JD time representation, equivalent
+            to about 20 picosecs.
+        """
+        if not isinstance(other, Time):
```
I'm wondering whether the checks are necessary, since ideally the comparison just breaks. Indeed,
```
In [20]: TimeDelta(1*u.day) < Time.now()                                                              
TypeError: '<' not supported between instances of 'TimeDelta' and 'Time'
```

But very strangely,
```
In [21]: TimeDelta(1*u.day, scale='tai') < Time('J2000')                                                                            
Out[21]: True
```
It looks like it has something to do with the specific scale picked - I'll investigate.

Ignoring that, would it be an idea to remove the checks? It might allow `Time` and `TimeDelta` can share the implementation.

mhvk replied:
I guess it is the usual "look before you leap" vs "easier to ask for forgiveness than permission" - https://docs.python.org/3/glossary.html#term-eafp

Generally, python seems to prefer the latter, and I think it is not illogical here either. E.g., suppose someone writes a different `MyTime` class (say, one that is integer + fraction in seconds), which they try to ensure can interact with `Time`, e.g., by defining subtraction with astropy's `Time` (which they can make work, because python will give their `__rsub__` a chance after our `__sub__` fails). If this also returns something that can have its absolute value taken and be compared with a quantity or `TimeDelta` (as they might well, since those are methods we've had for a long time), then the actual `isclose()` code will "just work". But in the current implementation, the `isinstance` check makes it impossible to ever work. (This is similar to numpy having `np.asarray` sprinkled throughout their code, making it not work with `Quantity`, and indeed `np.abs` sort-of working rather than just plain failing).

p.s. In principle, in EAFP, one does try/except and gives perhaps more informative error messages, but I've slowly moved to a state where I think one should just count on the user looking at the docstring when an error is raised. And in this case, one would generally get a quite informative `TypeError` from the subtraction anyway.

taldcroft replied:
I see your point and it is valid. OTOH, it happens frequently (based on real astropy issues I've seen) that users get confused because of an obscure error message raised in an unrelated bit of code when they pass a bad input. So there is a balance between addressing a problem that it almost sure to occur (bad inputs) vs. a new `MyTime` class that is less likely.

It's also worth noting that a major theme of recent Python development is type annotation. I've actually started paying some attention to this since there must be a reason the core devs are spending so much time on it. From this perspective we would annotate the input with *something*. Agreed that I'm not sure how this really works with duck-typing, which is certainly an incredibly useful concept.

I guess the middle ground is putting the main bit of the function in a try / except TypeError and re-raising with a better exception.

mhvk replied:
Agreed that the `MyTime` class is not very likely, and I should add that this is not a big deal! But if one does it without anything, the type of exception one gets is already quite clear:
```
In [11]: class TTime(Time): 
    ...:     def isclose(self, other, atol=1*u.ns): 
    ...:         return abs(self - other)<atol 
    ...:               
    ...:                                                                                               

In [12]: t = TTime.now()                                                                               

In [13]: dt = TimeDelta(100, format='jd')                                                              

In [14]: t.isclose('J2000')                                                                            
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-14-15f66b730fe3> in <module>
----> 1 t.isclose('J2000')

<ipython-input-11-b8b243443d6c> in isclose(self, other, atol)
      1 class TTime(Time):
      2     def isclose(self, other, atol=1*u.ns):
----> 3         return abs(self - other)<atol
      4 
      5 

TypeError: unsupported operand type(s) for -: 'TTime' and 'str'

In [15]: t.isclose(dt)                                                                                 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-f21ef40edf4f> in <module>
----> 1 t.isclose(dt)

<ipython-input-11-b8b243443d6c> in isclose(self, other, atol)
      1 class TTime(Time):
      2     def isclose(self, other, atol=1*u.ns):
----> 3         return abs(self - other)<atol
      4 
      5 

TypeError: bad operand type for abs(): 'TTime'
```

mhvk replied:
p.s. The annotations are indeed intriguing. I think currently they are only to help writing correct code, not execution, but I think it is a good way to make the intent clearer. On annotations, our very own `quantity_input` decorator is an exception in that it actually uses annotations at runtime. Actually, that is seeing some nice discussion over at #10156 with a draft new implementation at #10655 - the latter might even get to your long-standing suggestion of having a way to ask for Quantity input that gives the function plain arrays in the right units! Would be good to get your input!

mhvk commented on astropy/time/core.py:
```diff
@@ -2460,6 +2488,32 @@ def _make_value_equivalent(self, item, value):
+            Tolerance for equality with units of time (e.g. ``u.s`` or ``u.day``).
+            Default is one bit in the 128-bit JD time representation, equivalent
+            to about 20 picosecs.
+        """
```
Possibly contrary to my argument above, `TimeDelta` perhaps should have a different implementation since `rtol` does make sense for `TimeDelta`.

mhvk commented on astropy/time/core.py:
```diff
@@ -1180,6 +1180,36 @@ def __setitem__(self, item, value):
+            Time object for comparison.
+        atol : `~astropy.units.Quantity` or `~astropy.time.TimeDelta`
+            Absoute tolerance for equality with units of time (e.g. ``u.s`` or
+            ``u.day``). Default is one bit in the 128-bit JD time representation,
```
I wonder a bit about the default: when we were doing the hypothesis testing, I found that for UTC I had to allow 2 bits (since taking the difference with a UTC time means one has to do a scale change): 
https://github.com/astropy/astropy/blob/f56d8abf5dd39671c695e1e295293ef1490e2485/astropy/time/tests/test_precision.py#L502-L512

Maybe we should do that here too? (Also, I guess in principle, one can just put the default in in the initializer, rather than make it `None`.)

taldcroft replied:
I made the change.

About putting the atol in the signature, I thought about this and am on the edge, but somehow seeing a somewhat non-trivial expression in the function signature feels wrong: 
```
def isclose(self, other, atol=2 * np.finfo(float).eps * u.day):
```
I think the repr of `<Quantity 4.4408921e-16 d>` won't really be that helpful in the method docstring.

mhvk replied:
I'd definitely leave it in `ns` if we put it in the signature... But, yes, clearly a user would have to read the docstring anyway, so really it doesn't matter - let's stick with what you have.

MaxNoe commented:
Is isclose a ufunc? If yes, does it make sense to create a WIP branch adding all the ufuncs we need that different people can contribute?


I now see at least three issues / PRs that started implememting `Time.__array_ufunc__`, maybe we should bundle the effort, work in a separste branch until all ufincs we want are implemented and then make a PR from that branch.

I usually don't like longer lived dev branches, but in this case it might be appropriate.

mhvk commented:
:+1: to a separate branch to work on `ufunc` support! You may want to take a look at `quantity_helpers` to see how it was done for units (since eventually we'll want to support functions like `isclose` too...).

```

In [3]: type(np.isclose)
Out[3]: function

In [4]: type(np.isnat)
Out[4]: numpy.ufunc
```

taldcroft commented:
> since eventually we'll want to support functions like isclose too...

I still don't see that `np.isclose` should be supported. 

For `Time`, the `np.isclose` arg `rtol` literally makes no sense, and `equal_nan` at best requires a stretch to interpret "NaN" as missing values. Furthermore the default for `atol=1e-8` is not appropriate since it is a float and we require a `Quantity` or `TimeDelta`.

MaxNoe commented:
> For Time, the np.isclose arg rtol literally makes no sense

For calendar dates i would agree. For timescales base on linear time offsets from an epoch, I think rtol would be well-defined. (Unix, TT, TAI, ...)

taldcroft commented:
Sure, for unix time there is technically a way to define `rtol`. But in practice this would become quite dodgy since those relative values don't have units that are programmatically specified. In other words the `TimeUnix` class does not know that the unix representation is in `u.s`.

Perhaps more importantly, I cannot see any practical use for `rtol` even with unix time. Do you have a use case?

MaxNoe commented:
@taldcroft no, no really. Maybe unit testing of arithmetic with Time / TimeDelta?

taldcroft commented:
@mhvk - I think I have addressed all the points, including a compromise on duck-typing the input.  The code that relies on `other` is now in a try/except with a reasonable error message and some care taken to make it easier for the user to see what went wrong.

taldcroft commented:
> Maybe unit testing of arithmetic with Time / TimeDelta?

That testing is already solid. I think that using `rtol` would only degrade the testing by introducing additional multiplies done at 64-bit resolution (vs. 128 everywhere else).

MaxNoe commented:
> That testing is already solid. I think that using rtol would only degrade the testing by introducing additional multiplies done at 64-bit resolution (vs. 128 everywhere else).

I did not mean unit testing inside astropy but of packages using astropy, but you are probably right that there is no real use case here.

mhvk commented:
@MaxNoe - I'm a bit more with you on this, that `np.isclose(t1, t2)` might as well do something sensible (i.e., with `rtol=0` assumed). But where it makes most sense is for `TimeDelta`: `np.isclose(dt1, dt2, atol=...*u.ns, rtol=1e-9)` is reasonable and would have the benefit that it doesn't matter if `dt1/2` are quantities and `TimeDelta`.

Anyway, the method here is good and discoverable and definitely seems the best way to go for `Time` itself (@taldcroft - I'll review later today).

mhvk commented on astropy/time/core.py:
```diff
@@ -914,6 +914,45 @@ def __setitem__(self, item, value):
+            # to do a scale change.
+            atol = 2 * np.finfo(float).eps * u.day
+
+        if not isinstance(atol, (u.Quantity, TimeDelta)):
```
Is this still necessary given your (very nice) error message below?

mhvk commented on astropy/time/core.py:
```diff
@@ -2457,6 +2496,40 @@ def _make_value_equivalent(self, item, value):
+        """
+        try:
+            other_day = other.to_value(u.day)
+        except Exception as err:
```
I'd follow `Time.isclose()` and do `atol_day = atol.to_value(u.day)` inside here (and adjust the error message), with the `atol is None` statement moved in front.

mhvk reviewed (approved):
Looks very nice. Two nitpicks, but OK to ignore, so approving.

taldcroft commented:
@mhvk - thanks for the review. I thought about your comments for a bit and in the end I think for `atol` it is reasonable to be strict about the type. I think it would be unusual for a program using astropy to find it inconvenient to specify `atol` as a quantity.

In general I still lean toward guaranteeing correctness with tests that run on the advertised API data types, rather than hoping for correctness based only on not generating an exception. Understood this is a big fuzzy line that we'll continually be straddling.
